Encryption.md

```diff
@@ -75,55 +75,126 @@ written to the output stream.
+supports two levels of integrity verification (authentication): of the data (default), and 
 of the data combined with an optional AAD (“additional authenticated data”). The default 
 authentication allows to make sure the data has not been tampered with. An AAD is a free 
 text to be signed, together with the data. The user can, for example, pass the file name 
-with its version (or creation timestamp) as the AAD, to verify the file has not been 
-replaced with an older version.
+with its version (or creation timestamp) as the AAD, to verify that the file has not been 
+replaced with an older version (see below).
 
-Sometimes, a hardware acceleration of AES is unavialable (e.g. in Java 8). Then AES crypto 
+Sometimes, a hardware acceleration of AES is unavailable (e.g. in Java 8). Then AES crypto 
 operations are implemented in software, and can be somewhat slow, becoming a performance 
 bottleneck in certain workloads. AES-CTR is a regular (not authenticated) cipher.
 It is faster than AES-GCM, since it doesn’t perform integrity verification and doesn’t 
-calculate the authentication tag. For applications running without AES acceleration and 
+calculate the authentication tag. Actually, AES-GCM is a combination of AES-CTR and an 
+authentication module called GMAC. For applications running without AES acceleration and 
 willing to compromise on content verification, AES-CTR can provide a boost in Parquet 
 write/read throughput. The second Parquet algorithm encrypts the data content (pages) 
 with AES-CTR, and the metadata (Thrift structures) with AES-GCM. This allows to encrypt/decrypt 
 the bulk of the data faster, while still verifying the metadata integrity and making sure 
-the file had not been replaced with a wrong version. However, tampering with the page data 
+the file has not been replaced with a wrong version. However, tampering with the page data 
 might go unnoticed. 
 
-The `AesGcmV1` and `AesGcmCtrV1` structures contain an optional `aad_metadata` field that can 
-be used by a reader to retrieve the AAD string used for file encryption. The maximal allowed
-length of `aad_metadata` is 512 bytes.
-
-Parquet-mr/-cpp implementations use the RBG-based IV construction as defined in the NIST 
-SP 800-38D document for the GCM ciphers (section 8.2.2).
+The initialization vectors (IVs) in AES GCM and CTR ciphers must be unique for each encryption key, 
+since a single re-use of an IV breaks the encryption. An IV in these two modes is often called a 
+nonce (“number used once”). Parquet encryption uses the RBG-based (random bit generator) IV 
+construction as defined in the NIST SP 800-38D document for the GCM ciphers (section 8.2.2). 
 
 
 ### AES_GCM_V1
-All modules are encrypted with the AES-GCM cipher. The authentication tags (16 bytes) are 
-written after each ciphertext. The IVs (12 bytes) are written before each ciphertext.
+All modules are encrypted by an AES-GCM cipher, without padding. Unique IVs are generated and 
```

Comment:
Is the padding statement clear enough? Is there a reference we could point to that shows how to encrypt the last, under-filled block?

Reply:
I'm using the standard APIs for that, such as 
 `javax.crypto.Cipher.getInstance("AES/GCM/NoPadding");` in Java.
As for a reference, we can point to
https://crypto.stackexchange.com/questions/42412/gcm-padding-or-not
or
https://docs.oracle.com/javase/9/docs/api/javax/crypto/Cipher.html
or some other reference

Reply:
The mode used for a Java library is not sufficient to document this. A reference for exactly what that mode does would be, if not a description here of how this is done.

Reply:
Yep, makes sense. I'll try to find one, or let me know if you come across something suitable.

Reply:
Ok, it looks like the most authoritative references for this are 
https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38a.pdf
https://tools.ietf.org/html/rfc3686

CTR is explicitly excluded from the list of padded ciphers in the former. The latter says:
"AES-CTR does not require the plaintext to be padded to a  multiple of the block size"

Also, we have these from different sources:

"because CTR mode (contrary to CBC mode) does not require padding to block boundaries, and AES-CTR would also be compatible with no padding at all, or with strict PKCS#5 padding"

"Internally GCM really is CTR mode along with a polynomial hashing function applied on the ciphertext. CTR-mode doesn't need padding because you can just partly use the bits the last counter block generated and the polynomial hash does use (zero-)padding."

"GCM is a streaming mode which means that the ciphertext is only as long as the plaintext (not including authentication tag). GCM doesn't require a padding. This means that the PKCS5Padding version is actually only a synonym for NoPadding for convenience during programming. Some providers don't have this strange mode."

So, while the NIST spec text is not great on this, the picture is clear. GCM/NoPadding implementations will be compatible in any language and crypto provider. Same is true for CTR/NoPadding.

Reply:
I'm sure the modes work without padding.

The problem is that an implementer needs to know how to take the AES-GCM block cipher and produce ciphertext for the last block. The pseudo-code is something like this:

```java
val last_key_block = ...
val last_ciphertext_block = new byte[last_block.length]
for (i in range(0, last_block.length)) {
  last_ciphertext_block[i] = last_key_block[i] ^ last_block[i];
}
```

Reply:
The details of AES GCM and CTR ciphers are defined in the two NIST specs, 800-38D and 800-38A.
I believe an implementer doesn't need to know how the block ciphers are created. For all practical purposes, its sufficient for a Parquet encryption implementer to pick a provider that implements these two ciphers, with a 'no padding' option, according to the NIST specs. Using JCE, OpenSSL, BountyCastle or any other provider worth mentioning, would guarantee the parquet encryption interop.
Having said that - I agree the padding part is not verbalized very well in the NIST specs. If you feel its important we add these details to the spec, to remove any doubts about the requirements on  a provider implementation - lets do it.

Reply:
Are there "no padding" modes for C++, Python, etc.? The libraries you mention are all JVM-based.

Reply:
OpenSSL is C++, we're using it in the parquet-cpp implementation. Tested for interop with the parquet-mr.
Anyways, adding the no-padding details to the spec.

Reply:
Starting to look at this, I've realized I'm not comfortable with providing details that are not provided in the NIST specifications for these ciphers. The reason NIST explicitly ignores padding for the CTR and GCM modes is likely to be this: _".. CTR modes do not require any special measures to handle messages whose lengths are not multiples of the block size, since the modes work by XORing the plaintext with the output of the block cipher"._
Therefore, I think we need to require from a Parquet implementer to use a cryptographic provider that implements these two ciphers according to the NIST specification. That will be sufficient to guarantee an interop. As an example, we already know the SunJCE and OpenSSL providers are interoperable, since we have tested this in the parquet-mr and parquet-cpp implementations.

Reply:
Okay, then let's leave it at "no padding" then.